### PR TITLE
tree: handle schema contravariantly when generating input types 

### PR DESCRIPTION
**Pull request by CraigMacomber**
## Description

Handle schema contravariantly when generating input types.

## Breaking Changes

No runtime changes, but typing tweaks could break users on non-exact schema, and the enum schema APIs have been tweaked to avoid relying on non-exact schema, see changeset.

## Reviewer Guidance

The review process is outlined on [this wiki page](https://github.com/microsoft/FluidFramework/wiki/PR-Guidelines#guidelines).

### CODE REVIEW

File: packages/dds/tree/src/simple-tree/schemaTypes.ts [hunk 1]
```diff
@@ -487,16 +528,73 @@ export type TreeNodeFromImplicitAllowedTypes<
...
+ * Future versions of the API can relax these requirements.
+ *
+ * For example if {@link InsertableField} produced `Input<never>` for some union of schema (ex: `A | OptionalField<A>`),
+ * a future version could instead return a more flexible but still safe type, like `Input<A>`.
+ *
+ * Note that since `Input` is just a genic type, it might not always appear in contexts which simplify or inline types.
+ * Thus "Input" in mainly useful as an annotation visible in type signatures and documentation.
+ *
+ * @privateRemarks
+ * TODO: Actually use this.
```
Comment (CraigMacomber) [inline packages/dds/tree/src/simple-tree/schemaTypes.ts:540]: I could use some advice on how to associate these docs with the relevant APIs.

Wrapping the output is a bit messy. I have `@see Input` docs, but with Input not being package exported the build is passing, which seems wrong. Maybe (unlike in intelisense) API-Extractor expects a link tag there?

Maybe an `Insertable` namespace with a doc comment would be viable, at least for the public ones?

Comment (Josmithr) [reply packages/dds/tree/src/simple-tree/schemaTypes.ts:540]: Left another comment before seeing this one. I'm surprised intellisense treats `@see` like a link - TSDoc does not, so you do need the `{@link}` tag.

Comment (Josmithr) [reply packages/dds/tree/src/simple-tree/schemaTypes.ts:540]: I like the namespace idea in theory, but I have a suspicion that using one will break API-Extractor in the `fluid-framework` package ☹️

https://github.com/microsoft/rushstack/issues/4807

Would it be problematic to just package-export this type as `@system` for no other reason than documentation?

Comment (CraigMacomber) [reply packages/dds/tree/src/simple-tree/schemaTypes.ts:534]: > Would it be problematic to just package-export this type as @system for no other reason than documentation?

No, I can do that. It seems a bit hacky to create linkable docs by exporting unused types just for their docs, but there's nothing inherently bad about it, and we can do something else layer since its a system type.

I guess I'll do that.

File: .changeset/shy-dots-prove.md [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+const schema = [A, B];
+
+const config = new TreeViewConfiguration({ schema });
+const view = sharedTree.viewWith(config);
+
+// Does not compile since setter for root is typed `never` due to imprecise schema.
+view.root = [];
+```
+
+This is disallowed since the same schema type could be produced either of:
```
Comment (tylerbutler) [inline .changeset/shy-dots-prove.md:28]:
```suggestion
This is disallowed since the same schema type could produce either of:
```

**Review by tylerbutler — APPROVED** [review]
Just some nits. Approved for docs.

File: .changeset/shy-dots-prove.md (hunk 2 above)
Comment (CraigMacomber) [reply .changeset/shy-dots-prove.md:28]:
```suggestion
The assignment of `view.root` is disallowed since the same schema type could be either of:
```

File: .changeset/shy-dots-prove.md [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+class B extends schemaFactory.array("B", schemaFactory.number) {}
+
+// Gives imprecise type (typeof A | typeof B)[]
+const schema = [A, B];
+
+const config = new TreeViewConfiguration({ schema });
+const view = sharedTree.viewWith(config);
+
+// Does not compile since setter for root is typed `never` due to imprecise schema.
+view.root = [];
```
Comment (Josmithr) [inline .changeset/shy-dots-prove.md:25]: Curious: what does the error look like in this case? Is it intelligible?

File: packages/dds/tree/api-report/tree.alpha.api.md [hunk 4]
```diff
@@ -20,12 +26,17 @@ export type AllowedTypes = readonly LazyItem<TreeNodeSchema>[];
 type AllowedTypesUnsafe = readonly LazyItem<TreeNodeSchemaUnsafe>[];
 
 // @public
-type ApplyKind<T, Kind extends FieldKind, DefaultsAreOptional extends boolean> = {
+type ApplyKind<T, Kind extends FieldKind> = {
```
Comment (Josmithr) [inline packages/dds/tree/api-report/tree.alpha.api.md:29]: This is a breaking change, no?

File: packages/dds/tree/src/simple-tree/schemaTypes.ts [hunk 5]
```diff
@@ -40,8 +45,16 @@ export function isTreeNodeSchemaClass<
  * Types for use in fields.
  * @remarks
  * Type constraint used in schema declaration APIs.
+ *
+ * The order of types in the array is not significant.
+ * Additionally it is legal for users of this type to have the runtime and compile time order of items withing this array not match.
```
Comment (Josmithr) [inline packages/dds/tree/src/simple-tree/schemaTypes.ts:50]:
```suggestion
 * Additionally, it is legal for users of this type to have the runtime and compile time order of items within this array not match.
```

File: packages/dds/tree/src/simple-tree/schemaTypes.ts [hunk 6]
```diff
@@ -40,8 +45,16 @@ export function isTreeNodeSchemaClass<
...
+ * Therefor to ensure type safety, this these arrays should not be indexed, and instead just be iterated.
```
Comment (Josmithr) [inline packages/dds/tree/src/simple-tree/schemaTypes.ts:51]:
```suggestion
 * Therefor to ensure type safety, these arrays should not be indexed, and instead just be iterated.
```

File: packages/dds/tree/src/simple-tree/schemaTypes.ts [hunk 7]
```diff
@@ -40,8 +45,16 @@ export function isTreeNodeSchemaClass<
...
+ *
+ * Ideally this restriction would be modeled in the type itself, but it is not ergonomic to do so as there is no easy (when compared to arrays)
+ * way to declare and manipulate unordered sets of types in TypeScript.
+ *
  * Not intended for direct use outside of package.
- * @public
+ * @system @public
```
Comment (Josmithr) [inline packages/dds/tree/src/simple-tree/schemaTypes.ts:57]: This is also technically breaking. How confident are we that customers are not directly referencing this type in some way?

File: packages/dds/tree/src/simple-tree/schemaTypes.ts [hunk 8]
```diff
@@ -485,16 +523,92 @@ export type TreeNodeFromImplicitAllowedTypes<
...
 		: unknown;
 
+/**
+ * This type exists only to be linked from documentation to provide a single linkable place to document some details of
+ * "Input" types and how they handle schema.
+ *
+ * When a schema is used to describe data which is an input into an API, the API is [contravariant](https://en.wikipedia.org/wiki/Covariance_and_contravariance_(computer_science)) over the schema.
+ * (See also, [TypeScript Variance Annotations](https://www.typescriptlang.org/docs/handbook/2/generics.html#variance-annotations)).
+ *
+ * Since these schema are expressed using TypeScript types, its possible for the user of the API to provide non-exact values of these types which has implications that depended ont eh variance.
```
Comment (Josmithr) [inline packages/dds/tree/src/simple-tree/schemaTypes.ts:533]:
```suggestion
 * Since these schema are expressed using TypeScript types, it is possible for the user of the API to provide non-exact values of these types which has implications that depended on the variance.
```

File: packages/dds/tree/src/simple-tree/schemaTypes.ts [hunk 9]
```diff
@@ -485,16 +523,92 @@ export type TreeNodeFromImplicitAllowedTypes<
...
+ *
+ * Consider a field with schema type of `A | B` (where A and B are types of schema).
+ *
+ * - Reading the field behaves covariantly so {@link NodeFromSchema} of `<A | B>` is the same as `NodeFromSchema<A> | NodeFromSchema<B>`, indicating that either type of node can be read from the field.
+ * - Writing to the field behaves contravariantly. Since its unknown if the node actually has a schema `A` or a schema `B`, the only legal values (known to be in schema regardless of which schema the underlying node has) are values which are legal for both `A & B`.
```
Comment (Josmithr) [inline packages/dds/tree/src/simple-tree/schemaTypes.ts:538]:
```suggestion
 * - Writing to the field behaves contravariantly. Since it is unknown if the node actually has a schema `A` or a schema `B`, the only legal values (known to be in schema regardless of which schema the underlying node has) are values which are legal for both `A & B`.
```

File: packages/dds/tree/src/simple-tree/schemaTypes.ts [hunk 10]
```diff
@@ -485,16 +523,92 @@ export type TreeNodeFromImplicitAllowedTypes<
...
+ *
+ * Note that this is distinct from the case where the schema is `[A, B]`.
+ * In this case is it known that the field allows both A and B (the field can be set to an A or a B value).
```
Comment (Josmithr) [inline packages/dds/tree/src/simple-tree/schemaTypes.ts:541]:
```suggestion
 * In this case it is known that the field allows both A and B (the field can be set to an A or a B value).
```

File: packages/dds/tree/src/simple-tree/schemaTypes.ts [hunk 11]
```diff
@@ -485,16 +523,92 @@ export type TreeNodeFromImplicitAllowedTypes<
...
+ * When `A | B` is used, the field might allow
+ * A but not B (so assigning a B value would be out of schema),
+ * B but not A (so assigning an A value would be out of schema)
+ * or both A and B.
+ *
+ * This gets more extreme when given completely unspecified schema.
+ * For example if a field is just provided {@link ImplicitFieldSchema}, nothing is known about the content of the field.
+ * This means reading the field (via {@link TreeFieldFromImplicitField}) can give any valid tree field content,
```
Comment (Josmithr) [inline packages/dds/tree/src/simple-tree/schemaTypes.ts:549]: Nit
```suggestion
 * This means that reading the field (via {@link TreeFieldFromImplicitField}) can give any valid tree field content,
```

File: packages/dds/tree/src/simple-tree/schemaTypes.ts [hunk 12]
```diff
@@ -485,16 +523,92 @@ export type TreeNodeFromImplicitAllowedTypes<
...
+ * but there are no safe values which could be written to the field (since its unknown what values would be out of schema) so {@link InsertableTreeFieldFromImplicitField} gives `never`.
+ *
+ * To implement this variance correctly, computing of types for input and output have to use separate utilities
```
Comment (Josmithr) [inline packages/dds/tree/src/simple-tree/schemaTypes.ts:552]: Nit
```suggestion
 * To implement this variance correctly, the computation of types for input and output have to use separate utilities
```

File: packages/dds/tree/src/simple-tree/schemaTypes.ts [hunk 13]
```diff
@@ -485,16 +523,92 @@ export type TreeNodeFromImplicitAllowedTypes<
...
+ * which take very different approaches when encountering non-exact schema like unions or `ImplicitFieldSchema`.
+ * The utilities which behave contravariantly as required to handle input correctly link this documentation to indicate that this is how they behave.
+ *
+ * In addition to behaving contravariantly these input type computation utilities often have further limitations.
```
Comment (Josmithr) [inline packages/dds/tree/src/simple-tree/schemaTypes.ts:556]:
```suggestion
 * In addition to behaving contravariantly, these input type computation utilities often have further limitations.
```

File: packages/dds/tree/src/simple-tree/schemaTypes.ts [hunk 14]
```diff
@@ -485,16 +523,92 @@ export type TreeNodeFromImplicitAllowedTypes<
...
+ * This is due to TypeScript making it difficult to implement this contravariance exactly.
+ * When faced with these implementation limitations these contravariant type computation utilities error on the side of producing overly strict requirements.
+ * For example in the above case of `A | B`, the utilities might compute an allowed insertable type as `never` even if there happens to be a common value accepted by both `A` and `B`.
+ * Future versions of the API can relax these requirements as the type computations are made more accurate.
+ *
+ * For a more concrete example: if {@link InsertableTreeFieldFromImplicitField} produced `never` for a schema `A | OptionalField<A>`,
+ * a future version could instead return a more flexible but still safe type, like `A`.
+ *
+ * More generally: try and avoid providing non-exact schema, especially for the fields of other schema.
```
Comment (Josmithr) [inline packages/dds/tree/src/simple-tree/schemaTypes.ts:565]: Nit
```suggestion
 * More generally: try to avoid providing non-exact schema, especially for the fields of other schema.
```

File: packages/dds/tree/src/simple-tree/schemaTypes.ts [hunk 15]
```diff
@@ -485,16 +523,92 @@ export type TreeNodeFromImplicitAllowedTypes<
...
+ * While these APIs attempt to handle such cases correctly, there are limitations and known bugs in this handling.
+ * Code using non-exact schema is much more likely to have its compilation break due to updates of this package or even TypeScript,
+ * and thus compilation breaks to edge cases of non-exact schema handling, especially with recursive schema, are not considered breaking changes.
```
Comment (Josmithr) [inline packages/dds/tree/src/simple-tree/schemaTypes.ts:568]:
```suggestion
 * and thus compilation breaks due to edge cases of non-exact schema handling, especially with recursive schema, are not considered breaking changes.
```

File: packages/dds/tree/src/simple-tree/schemaTypes.ts [hunk 16]
```diff
@@ -485,16 +523,92 @@ export type TreeNodeFromImplicitAllowedTypes<
...
+ * This may change as the API become more stable.
+ *
+ * @privateRemarks
+ * There likely is a better way to share this documentation, but none was found at the time of writing.
```
Comment (Josmithr) [inline packages/dds/tree/src/simple-tree/schemaTypes.ts:572]: I wonder if there are JSDoc/TSDoc pre-processor tools out there that we could utilize to add custom behaviors like `{@include <some-target-docs-definition>}` and actually modify the doc contents during TSC's build. If the `d.ts` files then have properly TSDoc-compliant contents, we wouldn't have to worry about incompatibility with API-Extractor.

Alternatively, if one doesn't exist already, writing one probably wouldn't be all too difficult and might be a good FHL project.

File: packages/dds/tree/api-report/tree.alpha.api.md (hunk 4 above)
Comment (CraigMacomber) [reply packages/dds/tree/api-report/tree.alpha.api.md:29]: This is a system type. Would be real nice if that was shown in the API report. Can you fix that?

File: packages/dds/tree/src/simple-tree/schemaTypes.ts (hunk 16 above)
Comment (Josmithr) [reply packages/dds/tree/src/simple-tree/schemaTypes.ts:572]: Something like `include` would also be great for generating example code. I would love to check in examples as "test" code that we build and have them injected into our docs automagically.

Comment (Josmithr) [reply packages/dds/tree/src/simple-tree/schemaTypes.ts:572]: We could also use such a tool to generate blocks of contents for our existing custom tags like `@system` and `@legacy` so that end-user intellisense would include relevant notices.

File: packages/dds/tree/src/simple-tree/schemaTypes.ts [hunk 17]
```diff
@@ -485,16 +523,92 @@ export type TreeNodeFromImplicitAllowedTypes<
...
+ */
+export type Input<T extends never> = T;
+
 /**
  * Type of content that can be inserted into the tree for a node of the given schema.
+ *
+ * @see {@link Input}
+ *
+ * @typeparam TSchema - Schema to process.
+ * @typeparam TResult - Do not specify: default value used as implementation detail.
```
Comment (Josmithr) [inline packages/dds/tree/src/simple-tree/schemaTypes.ts:585]: I take it there is no reasonable way for us to hide these parameters that we strictly want to be inferred? I can't think of one, but you know TypeScript's generics _way_ better than I do.

File: packages/dds/tree/src/util/typeUtils.ts [hunk 18]
```diff
@@ -137,3 +137,90 @@ export type RestrictiveStringRecord<T> = {
...
+ * Converts a union type to a tuple type.
+ *
+ * @typeparam T - The union to convert.
+ * @typeparam A - Implementation detail: do not specify.
+ * @typeparam First - Implementation detail: do not specify.
+ *
+ * @privateRemarks
+ * `A` is a tuple to prepend members of the union to.
+ *
+ * https://www.hacklewayne.com/typescript-convert-union-to-tuple-array-yes-but-how and https://catchts.com/union-array both explain the general approach this uses pretty well.
```
Comment (Josmithr) [inline packages/dds/tree/src/util/typeUtils.ts:190]: Nit
```suggestion
 * {@link https://www.hacklewayne.com/typescript-convert-union-to-tuple-array-yes-but-how} and {@link https://catchts.com/union-array} both explain the general approach this uses pretty well.
```

File: packages/dds/tree/src/util/typeUtils.ts [hunk 19]
```diff
@@ -137,3 +137,90 @@ export type RestrictiveStringRecord<T> = {
...
+	A extends unknown[] = [],
+	First = PopUnion<Union>,
+> = IsUnion<Union> extends true
+	? UnionToTuple<Exclude<Union, First>, [First, ...A]>
+	: [Union, ...A];
+
+/**
+ * This is unsafe, meaning that the returned value might not match its type.
+ *
+ * For this the result to match its type, T must be a union of the types of each item in `items` in the order that they occur.
```
Comment (Josmithr) [inline packages/dds/tree/src/util/typeUtils.ts:205]:
```suggestion
 * For the result to match its type, T must be a union of the types of each item in `items` in the order that they occur.
```

**Review by Josmithr — COMMENTED** [review]
Left a handful of suggestions and questions. Biggest question: how confident are we that the type-wise breaking changes made here will not disrupt the majority of users?

File: .changeset/few-cobras-sing.md [hunk 20]
```diff
@@ -0,0 +1,21 @@
...
+---
+
+Add `.schema` member to alpha enum schema APIs
+
+The return value from `@alpha` APIs `enumFromStrings` and `adaptEnum` now has a property named `schema` which can be used to include it in a parent schema.
+This replaces the use of `typedObjectValues` which has been removed.
+
+Use of these APIs now look like:
+
+```typescript
```
[the remark this answers is not in the document]
Comment (CraigMacomber) [reply .changeset/few-cobras-sing.md:16]: Done.

File: packages/dds/tree/src/simple-tree/schemaTypes.ts (hunk 7 above)
Comment (CraigMacomber) [reply packages/dds/tree/src/simple-tree/schemaTypes.ts:57]: Since we have the comment "Not intended for direct use outside of package." no usage if this was possible if using as documented, so I think its ok. As that comment landed just before 2.0 (when system wasn't a well documented thing), I think this is fine.

File: packages/dds/tree/src/simple-tree/schemaTypes.ts (hunk 16 above)
Comment (CraigMacomber) [reply packages/dds/tree/src/simple-tree/schemaTypes.ts:572]: I'd love to have rust's doc tests, and ability to have file include macros (even in docs)...

I think making a "tag -> notice" injector as a build or publish step to match the API docs site would be the cleanest and most consistent. Would be great for system, legacy, deprecated, but could be used for other adhock cases like this.

Alternatively finding a way to link markdown files from /docs could also meet our needs here.

File: packages/dds/tree/api-report/tree.alpha.api.md (hunk 4 above)
Comment (Josmithr) [reply packages/dds/tree/api-report/tree.alpha.api.md:29]: I have a note to file a feature request. That's as far as I've gotten. But I suspect this would be an easy feature to get in.

Comment (Josmithr) [reply packages/dds/tree/api-report/tree.alpha.api.md:29]: I may spend some time this afternoon filing tickets and repros with them. It's been a while...

File: packages/dds/tree/src/simple-tree/schemaTypes.ts (hunk 17 above)
Comment (CraigMacomber) [reply packages/dds/tree/src/simple-tree/schemaTypes.ts:585]: Adding extra dummy parameters is just a way to have variables in these type computations which can be reused. I think its cleaner than making an additional "inner" type which we also would have to export. We could also just copy past the expression into every use site.

I don't find either of those workarounds better, but they are options we could take instead.

In this specific case though these docs are stale and there is no such parameter: oops!

File: packages/dds/tree/src/util/typeUtils.ts (hunk 18 above)
Comment (CraigMacomber) [reply packages/dds/tree/src/util/typeUtils.ts:190]: Is there a reason to use link tags in privateRemarks? Links are clickable in IntelliSense and inline without them, and they don't end up in the generated docs.

File: packages/dds/tree/package.json [hunk 21]
```diff
@@ -229,13 +229,15 @@
 	"typeValidation": {
 		"broken": {
 			"TypeAlias_InsertableTreeFieldFromImplicitField": {
```
[the remark this answers is not in the document]
Comment (CraigMacomber) [reply packages/dds/tree/package.json:231]: Type tests are basically meaningless for generic types since they just pass in `any`, and we don't actually define what using a schema of `any` is supposed to result in. There are real and important breaking changes that these do not catch, and non breaking changes (like adding a defaulted parameter you aren't allowed to use) which it considers breaking (as is the case for this one).

And no, these are not all system. This one is just public for example.

[the remark this answers is not in the document]
Comment (CraigMacomber) [reply packages/dds/tree/package.json:231]: Also, `@internal` types won't be included here since they are filtered based on release tag. We have some public type exported in the InternalTypes module, but those are not actually `@internal` just a flawed attempted to try and make systems types not get mixed in with the others.

**Comment by CraigMacomber** [comment]
> Left a handful of suggestions and questions. Biggest question: how confident are we that the type-wise breaking changes made here will not disrupt the majority of users?

It's hard to tell. It did break our in repo ai-collab code, but it's doing non-schema aware editing. I don't really know how many people are doing things that we never intended to support and are unsafe. I'll do a test build and try it in the examples repo.

File: .changeset/shy-dots-prove.md [hunk 22]
```diff
@@ -0,0 +1,59 @@
...
+const schema = [A, B] as const;
+const config = new TreeViewConfiguration({ schema });
+```
+
+```typescript
+const config = new TreeViewConfiguration({ schema: [A, B] });
+```
+
+To help update existing code which accidentally depended on this bug an `@alpha` API `unsafeArrayToTuple` has been provided.
+Many usages of this API will produce incorrectly types outputs, but when given `AllowedTypes` arrays which should not contain any unions, but accidentally got flatted to a single union, it can fix them:
```
Comment (alexvy86) [inline .changeset/shy-dots-prove.md]: Avoids the "A but B, but C, D" which was a bit hard to parse on first read.
```suggestion
Many usages of this API will produce incorrectly typed outputs. However, when given `AllowedTypes` arrays which should not contain any unions, but accidentally got flatted to a single union, it can fix them:
```

File: .changeset/shy-dots-prove.md [hunk 23]
```diff
@@ -0,0 +1,59 @@
...
+Improve strictness of input tree types when non-exact schema are provided
+
+When the type of the schema is not exactly specified, for example:
+
+```typescript
+const schemaFactory = new SchemaFactory("com.myApp");
+class A extends schemaFactory.object("A", {}) {}
+class B extends schemaFactory.array("B", schemaFactory.number) {}
+
+// Gives imprecise type (typeof A | typeof B)[]
```
Comment (alexvy86) [inline .changeset/shy-dots-prove.md:18]: I couldn't really understand why the suggestions to remove the ambiguity worked until I got to the end of the changeset and saw a comment pointing out this is the "correct" type, so pointing it out here seems good to make it clear from the start.
```suggestion
// Gives imprecise type (typeof A | typeof B)[], where [typeof A, typeof B] is what would work best for Tree
```

File: .changeset/shy-dots-prove.md (hunk 2 above)
Comment (alexvy86) [inline .changeset/shy-dots-prove.md]:
```suggestion
This is disallowed since the same schema type could be produced by either of:
```

**Review by alexvy86 — APPROVED** [review]
Didn't go through the functional changes, just looked at the changesets and left a couple of suggestions, but they otherwise look good. Definitely worth getting more approvals, particularly from API council given the breaking change.

File: packages/dds/tree/src/simple-tree/schemaTypes.ts (hunk 16 above)
Comment (Josmithr) [reply packages/dds/tree/src/simple-tree/schemaTypes.ts:572]: This can theoretically be done, but it's a bit fragile. E.g.
```typescript
/**
 * @see {@link https://github.com/microsoft/FluidFramework/blob/main/docs/README.md}
 */
```

File: packages/dds/tree/src/util/typeUtils.ts (hunk 18 above)
Comment (Josmithr) [reply packages/dds/tree/src/util/typeUtils.ts:190]: In theory so that refactoring tools update the text if the target changes.

**Review by Josmithr — APPROVED** [review]

File: packages/dds/tree/src/util/typeUtils.ts (hunk 18 above)
Comment (CraigMacomber) [reply packages/dds/tree/src/util/typeUtils.ts:190]: If you mean the target URL, I'm not sure how a refactoring tools would do that (no one is using a refactoring tool that updates a third party sit's URL scheme and our codebase at the same time), and if you mean the link display text, links in this format can't have that's not just the url so I don't think that applies here either.

Is there so other refactoring tool that I'm missing here?

Comment (Josmithr) [reply packages/dds/tree/src/util/typeUtils.ts:190]: Sorry, I missed that this one was a URL link. My comment applies to symbolic links in private remarks. For URLs, it probably doesn't really matter so long as intellisense handles them well without the tag.

**Comment by CraigMacomber** [comment]
> > Left a handful of suggestions and questions. Biggest question: how confident are we that the type-wise breaking changes made here will not disrupt the majority of users?
> 
> It's hard to tell. It did break our in repo ai-collab code, but it's doing non-schema aware editing. I don't really know how many people are doing things that we never intended to support and are unsafe. I'll do a test build and try it in the examples repo.

I have done the test integration into the examples repo. It has some preexisting problems, but I don't see anything new from these changes.
